src/curveToken.ts

```diff
@@ -0,0 +1,203 @@
+ * throw an error and a new solution will have to be found in that case
+ */
+async function getCurvePoolFromTokenAddress(address: string): Promise<string> {
+  // first try gauge data
+  const gaugeData = await getGauges();
+  if (gaugeData) {
+    const gaugeTokenInfo = Object.values(gaugeData).find(
+      ({ swap_token }) => swap_token === address
+    );
+
+    if (gaugeTokenInfo) {
+      return gaugeTokenInfo.swap;
+    }
+  }
+
+  // then factoryV2 pool data for metapools
+  const factoryV2PoolsData = await getFactoryV2Pools();
+  if (factoryV2PoolsData) {
+    const factoryV2PoolInfo = factoryV2PoolsData.find(
+      ({ address: _address }: { address: string }) => _address === address
+    );
+
+    if (factoryV2PoolInfo) return address;
+  }
+
+  const curveTokenAbi = await etherscanProvider.fetch("contract", {
+    action: "getabi",
+    address: address,
+    post: false,
+  });
+
+  const curveLpToken = new ethers.Contract(address, curveTokenAbi, provider);
+
+  if (
+    Object.keys(curveLpToken.functions).some((functionName) =>
+      functionName.startsWith("minter")
+    )
+  ) {
+    // if the token has a minter function, that is the original pool address
+    return await curveLpToken.minter();
+  }
+
+  throw new Error(`curve pool address: ${address} could not be found`);
+}
+
+export async function getCurveTokenInfo({
+  chainId,
+  address,
+  name,
+  symbol,
+  decimals,
+}: TokenInfo): Promise<CurveLpTokenInfo> {
+  const pool = await getCurvePoolFromTokenAddress(address);
+
+  const curvePoolAbi = await etherscanProvider.fetch("contract", {
+    action: "getabi",
+    address: pool,
+    post: false,
+  });
+
+  const curvePoolContract = new ethers.Contract(pool, curvePoolAbi, provider);
+
+  const addLiquidityFuncSig = Object.keys(curvePoolContract.functions).find(
+    (k) => k.startsWith("add_liquidity")
+  ) as string;
+  const removeLiquidityFuncSig = Object.keys(curvePoolContract.functions).find(
+    (k) => k.startsWith("remove_liquidity_one_coin")
+  ) as string;
+
+  /*
+   * We need this so that we can build our "poolAssets" definition which can be
+   * found in our curveLpToken type. We don't know ahead of time whether a given
+   * curve pool address has 2 or 3 tokens being traded on it and so we must
+   * validate this. Once knowing the number of coins in a given pool we can
+   * fetch those addresses and build the poolAssets array with the correct token
+   * addresses at the correct index
+   */
+  const numCoins = parseInt(
+    addLiquidityFuncSig.substring(
+      addLiquidityFuncSig.indexOf("[") + 1,
+      addLiquidityFuncSig.indexOf("]")
+    )
+  );
+
+  const curvePool = new ethers.Contract(
+    pool,
+    ["function coins(uint256) view returns (address)"],
+    provider
+  );
+
+  const coins = (await Promise.all(
+    [...Array(numCoins)].map((_, idx) => curvePool.coins(idx))
+  )) as [string, string] | [string, string, string];
```

Comment:
Hmm, I'd disagree, the array is always going to be length 2 or 3

Reply:
Yeah the issue though is that by casting to a union of tuples we lose a lot of the benefits of the type system. Tuple types work best when they fail at compile time if you try to accessing an element outside the set of known indices. 

```ts
const coins = ["a", "b"] as  [string, string] | [string, string, string];
const shouldBeACompileErrorButIsnt = coins[2];
```

We'll end up having to narrow this type everywhere through casting or typeguards, which will become error prone and just isn't worth it, so it's best we stick with `string[]` here 🙂

Reply:
Yeah we should make the change still. The array will never be empty as you point out, so that check isn't necessary.  The PR is fine to land w/ this change! 👍